File: Apple/server-driven-ui/server-driven-ui/library/definition/AppComponentDefinition.swift
//
//  AppComponentDefinition.swift
//  server-driven-ui
//
//  Created by Emma Vikström on 2026-04-10.
//

import Foundation


struct AppComponentDefinition: Codable, Identifiable{
    let name: Components
    let props: ComponentProps?
    let children: [AppComponentDefinition]?
    var key: String = UUID().uuidString
    
    var id: String { key }
    
    //Key genereras lokalt
    private enum CodingKeys: String, CodingKey {
        case name, props, children
    }
}


// Alla typer av komponeter som en AppComponentdefinition kan vara
enum Components: String, Codable{
    // Komponenter
    case ListComponent
    case ListSectionComponent
    case TextComponent
    case HeaderComponent
    case ButtonComponent
    case NavigationBarComponent
    case AccountCardComponent
    case DropDownAccountComponent
    case DropDownStringComponent
    case InputComponent
    case FormComponent
    case LocationPermissionComponent
    
    // Layouts
    case HorizontalLayout
    case VerticalLayout
    
    //Vyer
    case NavigationView
    case BankView
    case TransferBalanceView
    case NoDestinationView
    case RemoveAccountView
    case AddAccountView
    case LocationPermissionView
    case WeatherView
}

// Alla komponenters props (egenskaper)
struct ComponentProps: Codable{
    let title: String?
    let text: String?
    let id: Int?
    let action: String?
    let formAction: String?
    let pages: [Page]?
    let pageName: String?
    let destination: String?
    let accountNumber: String?
    let accountBalance: Double?
    let currency: String?
    let accountItems: [Account]?
    let stringItems: [String]?
    let sectionTitle: String?
}

File: Android/app/src/main/java/com/DT099G/exjobb/components/AppComponentDefinition.kt
package com.DT099G.exjobb.components

import kotlinx.serialization.Serializable

// --------- AppComponentDefinition --------
@Serializable
data class AppComponentDefinition(
    val name:Component,
    val props: ComponentProps? = null,
    val children: List<AppComponentDefinition>? = null,
    val key: String? = null
)


@Serializable
enum class Component{
    // Komponenter
    ListComponent,
    ListSectionComponent,
    TextComponent,
    HeaderComponent,
    ButtonComponent,
    NavigationBarComponent,
    AccountCardComponent,
    DropDownAccountComponent,
    DropDownStringComponent,
    InputComponent,
    FormComponent,
    LocationPermissionComponent,


    // Layouts
    HorizontalLayout,
    VerticalLayout,


    // Vyer
    NavigationView,
    BankView,
    NoDestinationView,
    TransferBalanceView,
    RemoveAccountView,
    AddAccountView,
    LocationPermissionView,
    WeatherView,
    NoAccessWeatherView
}

@Serializable
data class ComponentProps(
    var title: String? = null,
    var text: String? = null,
    var id: Int? = null,
    var action: String? = null,
    var formAction: String? = null,
    var pages: List<Page>? = null,
    var pageName: String? = null,
    var destination: String? = null,
    var accountNumber: String? = null,
    var accountBalance: Double? = null,
    var currency: String? = null,
    var accountItems: List<Account>? = null,
    var stringItems: List<String>? = null,
    var sectionTitle: String? = null

)
